src/msemu.c

```diff
@@ -58,106 +44,130 @@ unsigned char hex2bcd (unsigned char x)
-		int x = 19 - ms->lcd_cas;
-		// Use right half if necessary
-		if (lcdnum == LCD_R) x += 20;
+	srandom((unsigned int)time(NULL));
+	for (i = 0; i < 15; i++) {
+		do {
+			rnd = random();
+		} while (!isalnum(rnd));
+		*df_buf = rnd;
+		df_buf++;
+	}
 
-		// Write out all 8 bits to separate bytes, using the current emulated LCD color
-		int n;
-		for (n = 0; n < 8; n++)
-		{
-			int idx = n + (x * 8) + (newaddr * 320);
-			ms->lcd_datRGBA8888[idx] = ((val >> n) & 1 ? UI_LCD_PIXEL_ON : UI_LCD_PIXEL_OFF);
-		}
+	*df_buf = '-';
+}
 
-		// Let main loop know to update screen with new LCD data
-		ms->lcd_lastupdate = SDL_GetTicks();
-	} else {
-		log_debug(" * LCD%s W [ CAS] <- %02X\n",
-		  lcdnum == LCD_L ? "_L" : "_R", newaddr, val);
+/* Check if serial number in dataflash buffer is valid for Mailstation
+ *
+ * Returns 1 if number is valid, 0 if invalid.
+ *
+ * Disassembly of codeflash found that the verification process for serial num
+ * is that it starts at 0x7FFC8, is 16 bytes long, each character must be
+ * ASCII '0'-'9', 'A'-'Z', 'a'-'z', or '-'
+ * In all observed Mailstations, the last character is '-', but it does
+ * not seem to be enforced in Mailstation firmware; it is not enforced here
+ */
+static int ms_serial_valid(uint8_t *df_buf)
+{
+	int i;
+	int ret = MS_OK;
+
+	df_buf += DF_SN_OFFS;
 
-		// If CAS line is low, set current column instead
-		ms->lcd_cas = val;
+	for (i = 0; i < 16; i++) {
+		if (!isalnum(*df_buf) && *df_buf != '-') ret = MS_ERR;
+		df_buf++;
 	}
-}
 
+	return ret;
+}
 
 //----------------------------------------------------------------------------
 //
-//  Emulates reading from Mailstation LCD
+//  Resets Mailstation state
 //
-uint8_t readLCD(ms_ctx* ms, uint16_t newaddr, int lcdnum)
+void ms_power_on_reset(ms_ctx *ms)
 {
-	uint8_t *lcd_ptr;
-	uint8_t ret;
-
-	if (lcdnum == LCD_L) lcd_ptr = ms->lcd_dat1bit;
-	/* XXX: Fix the use of this magic number, replace with a const */
-	else lcd_ptr = &ms->lcd_dat1bit[4800];
-
-	/* XXX: This might need to be reworked to use non-viewable LCD memory */
-	// Wraps memory address if out of bounds
-	if (newaddr >= 240) newaddr %= 240;
-
-	// Check CAS bit on P2
-	if (ms->io[MISC2] & 8)
-	{
-		// Return data on currently selected LCD column
-		ret = lcd_ptr[newaddr + (ms->lcd_cas * 240)];
-	} else {
-		// Not sure what this normally returns when CAS bit low!
-		ret = ms->lcd_cas;
-	}
+	/* NOTE: While it might seem like it should be done, the keyboard
+	 * matrix, ms->key_matrix, should NOT be reset here. Since this is
+	 * set/cleared by UI functions as time goes on, clearing this for
+	 * a reset could lose keys that are being held down */
+
+	/* NOTE: The MS doesn't have a RAM clear at power on to the best of my
+	 * knowlegde. However, the emulation doesn't work quite right if RAM
+	 * is not cleared. This might be worth looking further in to at some
+	 * point. */
+
+	lcd_init(&ms->lcd_dat1bit, &ms->lcd_datRGBA8888, &ms->lcd_cas);
```

Comment:
If you F12 power off and F12 power on, the texture for the LCD screen will contain the last thing rendered before power off for a frame.

I think we can have `lcd_init` call `ui_update_lcd` after clearing the lcd buffers.

Reply:
I _thought_ I fixed this somewhere. I remember seeing it and it bothering me, and its not in my CWB where I am doing more cleanup. I'll take a closer look at this and drop a followup.

Reply:
Okay, well, something is fucky with my codebases because I was trying to move forward while this PR was open. I seem to have this both fixed and not fixed but have no idea where the actual fix is. I'm going to just merge this PR and update all of my weirdo branches and re-investigate this bug.